app/controllers/api/v2/compute_resources_controller.rb

```diff
@@ -238,11 +238,48 @@ def refresh_cache
         end
       end
 
+      api :GET, "/compute_resources/:id/available_virtual_machines/", N_("List available virtual machines for a compute resource")
+      param :id, :identifier, :required => true
+      def available_virtual_machines
+        @available_virtual_machines = @compute_resource.vms.all
+      end
+
+      api :GET, "/compute_resources/:id/available_virtual_machines/:vm_id", N_("Show a virtual machine")
+      param :id, :identifier, :required => true
+      param :vm_id, :identifier, :required => true
+      def show_vm
+        @vm = @compute_resource.find_vm_by_uuid(params[:vm_id])
+        attributes = @vm.attributes.deep_symbolize_keys
+        attributes.delete(:parent)
+        render :json => attributes
+      end
+
+      api :PUT, "/compute_resources/:id/available_virtual_machines/:vm_id/power", N_("Power a Virtual Machine")
+      param :id, :identifier, :required => true
+      param :vm_id, :identifier, :required => true
+      def power_vm
+        @vm = @compute_resource.vms.find { |vm| vm.id = params[:vm_id] }
+        action = @vm.ready? ? :stop : :start
+        begin
+          @vm.send(action)
+          render_message(_('%{vm} is now %{vm_state}') % {:vm => @vm, :vm_state => @vm.state.capitalize})
+        rescue Foreman::Exception => e
+          render_exception(e, :status => :unprocessable_entity)
+        end
+      end
+
+      api :DELETE, "/compute_resources/:id/available_virtual_machines/:vm_id", N_("Delete a Virtual Machine")
+      param :id, :identifier, :required => true
+      param :vm_id, :identifier, :required => true
+      def destroy_vm
```

Comment:
Returns no errors on failure:

```
[vagrant@foreman-devel ~]$curl -H "Accept: application/json" -H "Content-Type: application/json" -k -X DELETE -u admin:changeme https://$(hostname)/api/v2/compute_resources/1/available_virtual_machines/e22382a6-b89a-458a-bb7c-31dfe377c568 -d '{}'| json_reformat
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   350    0   348  100     2   1300      7 --:--:-- --:--:-- --:--:--  1303
{
    "id": 1,
    "name": "libvirt",
    "description": "",
    "url": "qemu+tcp://root@192.168.100.1/system",
    "user": null,
    "password": null,
    "uuid": "C4B5C0CC-1FED-11B2-A85C-F20F1972B825",
    "created_at": "2017-11-07T09:56:59.549Z",
    "updated_at": "2017-11-07T09:56:59.549Z",
    "attrs": {
        "display": "vnc",
        "setpw": 0
    },
    "email": null,
    "caching_enabled": true,
    "http_proxy_id": null,
    "domain": null
}
```

```
2020-01-02T08:54:51 [I|app|c4c74efd] Started DELETE "/api/v2/compute_resources/1/available_virtual_machines/e22382a6-b89a-458a-bb7c-31dfe377c568" for ::1 at 2020-01-02 08:54:51 +0000
2020-01-02T08:54:51 [I|app|c4c74efd] Processing by Api::V2::ComputeResourcesController#destroy_vm as JSON
2020-01-02T08:54:51 [I|app|c4c74efd]   Parameters: {"apiv"=>"v2", "id"=>"1", "vm_id"=>"e22382a6-b89a-458a-bb7c-31dfe377c568", "compute_resource"=>{}}
2020-01-02T08:54:51 [D|app|c4c74efd] Authenticated user admin against INTERNAL authentication source
2020-01-02T08:54:51 [D|app|c4c74efd] Post-login processing for admin
2020-01-02T08:54:51 [I|app|c4c74efd] Authorized user admin(Admin User)
2020-01-02T08:54:51 [D|app|c4c74efd] Post-login processing for admin
2020-01-02T08:54:51 [D|tax|c4c74efd] Current location set to none
2020-01-02T08:54:51 [D|tax|c4c74efd] Current organization set to none
2020-01-02T08:54:51 [D|tax|c4c74efd] Current location set to none
2020-01-02T08:54:51 [D|tax|c4c74efd] Current organization set to none
2020-01-02T08:54:51 [W|app|c4c74efd] Failed retrieving libvirt vm by uuid e22382a6-b89a-458a-bb7c-31dfe377c568
2020-01-02T08:54:51 [D|app|c4c74efd] Backtrace for 'Failed retrieving libvirt vm by uuid e22382a6-b89a-458a-bb7c-31dfe377c568' error (Libvirt::RetrieveError): Call to virDomainLookupByUUID failed: Domain not found: no domain with matching uuid 'e22382a6-b89a-458a-bb7c-31dfe377c568'
 | /home/vagrant/.rvm/gems/ruby-2.4.1/gems/fog-libvirt-0.7.0/lib/fog/libvirt/requests/compute/list_domains.rb:9:in `lookup_domain_by_uuid'
 | /home/vagrant/.rvm/gems/ruby-2.4.1/gems/fog-libvirt-0.7.0/lib/fog/libvirt/requests/compute/list_domains.rb:9:in `list_domains'
 | /home/vagrant/.rvm/gems/ruby-2.4.1/gems/fog-libvirt-0.7.0/lib/fog/libvirt/models/compute/servers.rb:15:in `get'
 | /home/vagrant/projects/foreman/app/models/compute_resource.rb:185:in `find_vm_by_uuid'
 | /home/vagrant/projects/foreman/app/models/compute_resources/foreman/model/libvirt.rb:43:in `find_vm_by_uuid'
 | /home/vagrant/projects/foreman/app/models/compute_resources/foreman/model/libvirt.rb:51:in `destroy_vm'
 | /home/vagrant/projects/foreman/app/controllers/api/v2/compute_resources_controller.rb:275:in `destroy_vm'

```

Reply:
That is strange, when I tried to delete a VM that does not exist I got this message:
{
  "error": {
    "message": "HTTP response code is 404. The response body is empty."
  }
}
I wonder why you did not get the same error

Reply:
I see now that this happens when trying to delete a libvirt VM. I will fix this.

Reply:
I discovered that this happens because whenever libvirt is trying to delete a VM that does not exist it does not return an error but returns true. you can see it in the libvirt.rb file:

    def destroy_vm(uuid, args = { })
      find_vm_by_uuid(uuid).destroy({ :destroy_volumes => true }.merge(args))
    rescue ActiveRecord::RecordNotFound
      true
    end

In other providers such as oVirt or Vmware the function looks the same but the error that is returned is not ActiveRecord::RecordNotFound therefore, they do not return true like in libvirt.

Do you think that should be changed?

Reply:
It should probably be fixed at some point. You can do it here if you think it is a small change or just create a new redmine issue.